### ISSUE TITLE
Fix duplication between default and bundled gems

### ISSUE THREAD

**Issue by bkuhlmann** [janlelis/stdgems#21]
## Overview

Hello. :wave: Thanks for maintaining the index of these gems. The JSON API is quite helpful and makes my life a lot easier.

I stumbled across a problem where if you combine the [Default API](https://stdgems.org/default_gems.json) with the [Bundled API](https://stdgems.org/bundled_gems.json), you end up with duplicate gem entries. Would it be possible to de-duplicate these APIs so there isn't double entry?

Another option, which might be nice, would be to supply a `index.json` API which could combine both APIs as one for convenience. At the moment, I merge them by making two API calls but would be handy to have a single API call.

## Screenshots/Screencasts

Here's an illustration of how the above plays out when wired up via my [Pennyworth](https://alchemists.io/projects/pennyworth) gem (planning to release support with Standard Gems later this week):

https://github.com/janlelis/stdgems/assets/50245/58be81fc-d591-4acf-89c8-bd8cd029620f

The above demonstration combines both the default and bundled APIs. I could de-duplicate these entries myself before presenting within the UI but was thinking the root issue might be work fixing first.

## Steps to Recreate

If you search through both APIs, you'll see duplicate entries for the Debug, Matrix, Did You Mean, etc gems.

**Comment by janlelis** [comment]
Hi @bkuhlmann, great that these JSON files were helpful for you. Pennyworth looks exciting!

Your example illustrates well why these duplicate entries can be problematic sometimes. Still, I'd like to keep both categories of gems separate, since their requirements (e.g. which properties per gem need to exist) differ slightly.

That's why I like your proposal of an (additional) unified JSON API. I'll give it a shot

**Comment by janlelis** [comment]
…although in the case of *debug*, it actually is a different gem. But that probably remains an exception...

**Comment by bkuhlmann** [comment]
Thanks. :bow: Yeah, definitely keep the two categories separate but having a complete index would be neat. I ended up releasing Pennyworth 15.1.0 with support for a _simulated_ index by making an API request for the bundled gems and then the default gems. Then I make the combined array unique, by gem name, which works out nicely.

I didn't realize the default Debug gem is different than the bundled Debug gem until you pointed that out. Hmm.

**Comment by janlelis** [comment]
Hey again, I have just released v4 with a combined json: https://github.com/janlelis/stdgems/blob/main/stdgems.json - I hope this simplifies things for you!

**Comment by bkuhlmann** [comment]
Thanks! This is great but I ran into two issues with the last issue being the most important:

- **Inconsistent Endpoints**: The endpoints are named as follows: `stdgems.json`, `bundled_gems.json`, and `default_gems.json`. Previously, I was able to use this format for the API calls: `<name>_gems.json`. However, with the introduction of `stdgems.json` I can no longer do that. The fix, on my end, was to use a hash map to translate all three endpoints into proper API calls. Not a big problem but might be nice to be consistent.
- **Inconsistent Data Structure**: Both the `bundled_gems.json` and `default_gems.json` endpoints use the same JSON API response of `gems: []` but with `stdgems.json` it only answers back a `gems: {}`. The difference between getting an *array* for two of the API responses and the third API response being a *hash* means I'm better off making two API calls which combine the bundled and default gems together (i.e. back to where I started). If the `stdgems.json` API was updated be identical in structure that would mean I could process all three API responses without any additional custom code on my side. 😅 

Would making the JSON API response identical in structure be possible? If so, then yes, this would be an awesome upgrade to the API and reduce a lot of code on my side. :bow:

**Comment by janlelis** [comment]
Oops, didn't mean to make `gems` a hash in the first place... It's fixed in the latest release. Thanks for bringing this up.

I'll keep the different endpoint naming scheme to highlight that the API still differs in some aspects (e.g. bundled/default namespace under versions) and that it is constructed from the other JSONs.

**Comment by bkuhlmann** [comment]
Thanks. The above fix works but I hit another inconsistency. This time with the version information:

- **Bundled/Default**: Always answers back an array of version hashes but only a single level deep. No categorization.
- **Stdgems**: Answers back a version hash but it's now nested with either a `bundle` or `default` prefix.

Is this intended behavior? I'm also confused by this because I never see this show up in the `stdgems.json`  response:

``` json
versions: {
  bundled: {1.2.3: "1.2.3"},
  default: {1.2.3: "1.2.3"}
}
```

Can all three API responses be consistent in that they only answer back a single array? Also I don't think you need to have nesting at all since the `currentType` key would give you what you want. Maybe I'm missing something?

**Comment by janlelis** [comment]
Some thoughts of why it's structured like this:

Some gems travel from bundled to default or vice versa. I think in many cases it's significant to know if the gem that came with a specific Ruby version was a default gem or a bundled one that's why I want to keep the version info separate (and even introduce this inconsistency).

One way of making the format consistent would be to introduce this nesting in the `{bundled,default}_gems.json` files, too - but I don't know if the breaking change in their format would be worth it.

I think of `stdgems.json` as a convenient additional (slightly different) interface to the data. I agree it is kind of surprising though and needs more documentation

**Comment by bkuhlmann** [comment]
Thanks. Hmm, couldn't you simplify this further by having one version format that looks something like this?

``` ruby
versions: [
  {short: "3.2", long: "3.2.0", type: "bundled"},
  {short: "3.1", long: "3.1.0", type: "default"},
  {short: "3.0", long: "3.0.0", type: "bundled"}
]
```

This would yield the following benefits:

- A chronological (descending order) of version history.
- The versions use keys which makes pattern matching easier.
- The `type` lets you know if the gem was *bundled* or *default* and at what version. Because of the chronology, you also know when it switched type and when.
- This could be applied across all APIs for improved consistency.

Always error on the side of consistency. Better for the maintainer and the consumer. :sweat_smile:

---

ℹ️ I released [Pennyworth 15.2.0](https://alchemists.io/projects/pennyworth) which uses the new [stdgems API](https://stdgems.org/stdgems.json). Works great and managed to clean up some code. I'm going to close this issue since it resolves my original request.
